

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>BEELINE &mdash; BEELINE  documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="BLRun package" href="BLRun.html" />
    <link rel="prev" title="Welcome to BEELINE’s documentation!" href="index.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> BEELINE
          

          
            
            <img src="_static/beeline.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">BEELINE</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="#getting-started">Getting Started</a></li>
<li class="toctree-l2"><a class="reference internal" href="#tutorial">Tutorial</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#project-outline">Project outline</a></li>
<li class="toctree-l3"><a class="reference internal" href="#input-datasets">Input Datasets</a></li>
<li class="toctree-l3"><a class="reference internal" href="#config-files">Config files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#running-the-pipeline">Running the pipeline</a></li>
<li class="toctree-l3"><a class="reference internal" href="#running-the-evaluation-scripts">Running the evaluation scripts</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#details-of-supported-algorithms">Details of supported algorithms</a></li>
<li class="toctree-l2"><a class="reference internal" href="#developer-guide">Developer Guide</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#adding-a-new-grn-inference-method">Adding a new GRN inference method</a></li>
<li class="toctree-l3"><a class="reference internal" href="#adding-a-new-evaluation-technique">Adding a new evaluation technique</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#api-reference">API Reference</a><ul>
<li class="toctree-l3"><a class="reference internal" href="BLRun.html">BLRun package</a></li>
<li class="toctree-l3"><a class="reference internal" href="BLEval.html">BLEval package</a></li>
<li class="toctree-l3"><a class="reference internal" href="BLPlot.html">BLPlot package</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BoolODE.html">BoolODE</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">BEELINE</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>BEELINE</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/BEELINE.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="beeline">
<span id="id1"></span><h1>BEELINE<a class="headerlink" href="#beeline" title="Permalink to this headline">¶</a></h1>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>BEELINE provides a set of tools for evaluating methods that infer gene
regulatory networks (GRN) from single-cell gene expression
data. The BEELINE framework is divided into the following modules:</p>
<ul class="simple">
<li><a class="reference internal" href="BLRun.html#blrun"><span class="std std-ref">BLRun package</span></a> : contains the BEELINE’s Runner module, a Python
wrapper for 12 GRN inference algorithms with options to add new
methods.</li>
<li><a class="reference internal" href="BLEval.html#bleval"><span class="std std-ref">BLEval package</span></a> : contains the BEELINE’s Evaluation module that
provides easy-to-use tools for evaluating GRN reconstructions.</li>
<li><a class="reference internal" href="BLPlot.html#blplot"><span class="std std-ref">BLPlot package</span></a> : contains the BEELINE’s plotting module for
visualizing output from BLEval.</li>
</ul>
</div>
<div class="section" id="getting-started">
<h2>Getting Started<a class="headerlink" href="#getting-started" title="Permalink to this headline">¶</a></h2>
<p>The BEELINE pipeline interfaces with the implementations of various
algorithms through Docker containers. BEELINE natively supports Debian-based distributions (e.g. Ubuntu), and also supports Windows/other Linux distributions. <a class="reference external" href="https://docs.docker.com/engine/install/">Install Docker Engine for your system.</a>
<br> If you prefer/your system supports a graphical interface, you may instead <a class="reference external" href="https://docs.docker.com/desktop/">install Docker Desktop.</a></p>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p>Setup docker to run docker without using sudo each time by adding yourself to the Docker group. You may have to restart your shell for the changes to take effect.</p>
<div class="code sh highlight-default notranslate"><div class="highlight"><pre><span></span>sudo usermod -aG docker $USER
</pre></div>
</div>
<p class="last">See more details <a class="reference external" href="https://askubuntu.com/questions/477551/how-can-i-use-docker-without-sudo">here</a>.</p>
</div>
<p>Once docker has been set up correctly, the next step is to create the
docker containers for each of the algorithms.  The script
<cite>initialize.sh</cite> builds these containers. Run the following in the
terminal</p>
<div class="code bash highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">.</span> <span class="n">initialize</span><span class="o">.</span><span class="n">sh</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This step will take a while! Ensure that your system also has ~20GB of memory available to build each image.</p>
</div>
<p>BEELINE requires <a class="reference external" href="https://www.anaconda.com/docs/getting-started/miniconda/install">Miniconda</a> (or Anaconda) for package management. Run the following command to automatically create an Anaconda virtual environment named BEELINE that contains all the necessary dependencies.</p>
<div class="code bash highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">.</span> <span class="n">setupAnacondaVENV</span><span class="o">.</span><span class="n">sh</span>
</pre>
</div>

    <div class="admonition tip">
    <p class="first admonition-title">Tip</p>
    <p>Anaconda isolates different sets of dependencies into each environment that it creates. To use an environment, you must "activate" it. You may have to reactivate it after logging back into a server or switching to another environment with</p>
    <div class="code sh highlight-default notranslate"><div class="highlight"><pre><span></span>conda activate BEELINE</pre></div>
    </div>
  </div>

  <p>BEELINE should now be installed and ready to run benchmarks.</p>
<p>To compute proposed reconstructions using the 12 GRN algorithms on the example dataset, run</p>
<div class="code python highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">BLRunner</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">config</span> <span class="n">config</span><span class="o">-</span><span class="n">files</span><span class="o">/</span><span class="n">config</span><span class="o">.</span><span class="n">yaml</span>
</pre></div>
</div>
<p>To compute areas under the ROC and PR curves for the proposed reconstructions, run</p>
<div class="code bash highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">BLEvaluator</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">config</span> <span class="n">config</span><span class="o">-</span><span class="n">files</span><span class="o">/</span><span class="n">config</span><span class="o">.</span><span class="n">yaml</span> <span class="o">--</span><span class="n">auc</span>
</pre></div>
</div>
<p>To display the complete list of
evaluation options, run</p>
<div class="code bash highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">BLEvaluator</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">help</span>
</pre></div>
</div>
</div>
<div class="section" id="tutorial">
<h2>Tutorial<a class="headerlink" href="#tutorial" title="Permalink to this headline">¶</a></h2>
<p>This tutorial will first explain the structure of the BEELINE repository,
with a walkthrough of the different components that the user can customize.</p>
<div class="section" id="project-outline">
<h3>Project outline<a class="headerlink" href="#project-outline" title="Permalink to this headline">¶</a></h3>
<p>The BEELINE repository is structured as follows:</p>
<div class="code text highlight-default notranslate"><div class="highlight"><pre><span></span>BEELINE
|-- inputs/
|   `-- examples/
|       `-- GSD/
|           |--refNetwork.csv
|           |--PseudoTime.csv
|           `--ExpressionData.csv
|-- config-files/
|   `-- config.yaml
|-- BLRun/
|   |-- sinceritiesRunner.py
|   `-- ...
|-- BLPlot/
|   |-- NetworkMotifs.py
|   `-- CuratedOverview.py
|-- BLEval/
|   |-- parseTime.py
|    `-- ...
|-- Algorithms/
|     |-- SINCERITIES/
|     `-- ...
|-- LICENSE
|-- initialize.sh
|-- BLRunner.py
|-- BLEvaluator.py
|-- README.md
`-- requirements.txt
</pre></div>
</div>
</div>
<div class="section" id="input-datasets">
<span id="id3"></span><h3>Input Datasets<a class="headerlink" href="#input-datasets" title="Permalink to this headline">¶</a></h3>
<p>The sample input data set provided is generated by <a class="reference internal" href="BoolODE.html#boolode"><span class="std std-ref">BoolODE</span></a>
using the Boolean model of <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/26573569">Gonadal Sex Determination</a> as input.  Note that
this dataset has been pre-processed to produce three files that are
required in the BEELINE pipline.</p>
<ol class="arabic simple">
<li><cite>ExpressionData.csv</cite> contains the RNAseq data, with genes as
rows and cell IDs as columns. This file is a required input to the
pipline. Here is a <a class="reference external" href="https://github.com/Murali-group/Beeline/blob/master/inputs/example/GSD/ExpressionData.csv">sample ExpressionData.csv file</a></li>
<li><cite>PseudoTime.csv</cite> contains the pseudotime values for the cells in
<cite>ExpressionData.csv</cite>.  We recommend using the Slingshot method to
obtain the pseudotime for a dataset. Many algorithms in the
pipeline require a pseudotime file as input. Here is a <a class="reference external" href="https://github.com/Murali-group/Beeline/blob/master/inputs/example/GSD/PseudoTime.csv">sample
PseudoTime file</a>.</li>
<li><cite>refNetwork.csv</cite> contains the ground truth network underlying the
interactions between genes in <cite>ExpressionData.csv</cite>. Typically this
network is not available, and will have to be curated from various
Transcription Factor databases. While this file is not a
requirement to run the base pipeline, a reference network is
required to run some of the performance evaluations in
<a class="reference internal" href="BLEval.html#bleval"><span class="std std-ref">BLEval package</span></a>.  Here is a <a class="reference external" href="https://github.com/Murali-group/Beeline/blob/master/inputs/example/GSD/refNetwork.csv">sample refNetwork.csv file</a></li>
</ol>
<p>The figure below shows the t-SNE visualization of the expression
data from the example dataset.</p>
<img alt="_images/SlingshotOutputVis.png" src="_images/SlingshotOutputVis.png" />
<p>This dataset shows a bifurcating
trajectory, as is evidenced by the part (a) of the figure, where
each ‘cell’ is colored by the timepoint at which it was sampled
in the simulation (the darker colors indicate earlier time points).
Clustering the simulation confirms the two trajectories, indicated
in red and blue respectively in part (b). Finally, running Slingshot
on this dataset and specifying the cluster of cells corresponding to
the early time points yields two pseudotime trajectories, shown in part (c).
For details on the generation of this simulated dataset, see <a class="reference internal" href="BoolODE.html#boolode"><span class="std std-ref">BoolODE</span></a>.</p>
<div class="admonition attention">
<p class="first admonition-title">Attention</p>
<p class="last">Please ensure that any input dataset you create is
comma separated, and contains the correct style of
column names.</p>
</div>
</div>
<div class="section" id="config-files">
<span id="configfiles"></span><h3>Config files<a class="headerlink" href="#config-files" title="Permalink to this headline">¶</a></h3>
<p>Beeline uses <a class="reference external" href="https://yaml.org/">YAML</a> files to allow users to
flexibly specify inputs and algorithm run parameters. A sample config
file is provided in <a class="reference external" href="https://github.com/Murali-group/Beeline/blob/master/config-files/config.yaml">here</a>. A
config file should contain at minimum</p>
<div class="code text highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">input_settings</span><span class="p">:</span>
    <span class="n">input_dir</span> <span class="p">:</span> <span class="s2">&quot;Base input directory name, relative to current working directory (recommended: inputs)&quot;</span>
    <span class="n">dataset_dir</span><span class="p">:</span> <span class="s2">&quot;Subdirectory of input_dir where the datasets are placed in&quot;</span>
    <span class="n">datasets</span><span class="p">:</span>
        <span class="o">-</span> <span class="n">name</span><span class="p">:</span> <span class="s2">&quot;Dataset name&quot;</span>
          <span class="n">exprData</span><span class="p">:</span> <span class="s2">&quot;Expression Data filename&quot;</span>
          <span class="n">cellData</span><span class="p">:</span> <span class="s2">&quot;PseudoTime filename&quot;</span>
          <span class="n">trueEdges</span><span class="p">:</span> <span class="s2">&quot;Ground truth network filename&quot;</span>
    <span class="n">algorithms</span><span class="p">:</span>
        <span class="o">-</span> <span class="n">name</span><span class="p">:</span> <span class="s2">&quot;Algorithm name&quot;</span>
          <span class="n">params</span><span class="p">:</span>
              <span class="c1"># Any other parameters that can be passed to</span>
              <span class="c1"># this algorithm</span>
              <span class="n">should_run</span><span class="p">:</span> <span class="p">[</span><span class="kc">True</span><span class="p">]</span> <span class="c1"># or False</span>

<span class="n">output_settings</span><span class="p">:</span>
    <span class="n">output_dir</span> <span class="p">:</span> <span class="s2">&quot;Base output directory name, relative to current working directory (recommended: outputs)&quot;</span>
    <span class="n">output_prefix</span><span class="p">:</span> <span class="s2">&quot;Prefix for writing evaluation summary files&quot;</span>
</pre></div>
</div>
<p>Apart from indicating the path to the base directory and the specific
folder containing the input, the config file indicates which
algorithms should be run, along with the parameters to be passed to
the algorithms, if any. For a list of parameters that the pipeline
currently passes to the algorithms, see  <a class="reference internal" href="#algorithms"><span class="std std-ref">Details of supported algorithms</span></a>.  Finally,
the YAML file also specifies the output_dir where the outputs are written.
Note that the output directory structure under output_dir is same
as the input directory strucutre under input_dir. For example, if the config file
contains the following:</p>
<div class="code text highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">input_settings</span><span class="p">:</span>
    <span class="n">input_dir</span> <span class="p">:</span> <span class="s2">&quot;inputs&quot;</span>
    <span class="n">dataset_dir</span><span class="p">:</span> <span class="s2">&quot;example&quot;</span>
    <span class="n">datasets</span><span class="p">:</span>
        <span class="o">-</span> <span class="n">name</span><span class="p">:</span> <span class="s2">&quot;GSD&quot;</span>
          <span class="n">exprData</span><span class="p">:</span> <span class="s2">&quot;ExpressionData.csv&quot;</span>
          <span class="n">cellData</span><span class="p">:</span> <span class="s2">&quot;PseudoTime.csv&quot;</span>
          <span class="n">trueEdges</span><span class="p">:</span> <span class="s2">&quot;refNetwork.csv&quot;</span>
    <span class="n">algorithms</span><span class="p">:</span>
        <span class="o">-</span> <span class="n">name</span><span class="p">:</span> <span class="s2">&quot;PIDC&quot;</span>
          <span class="n">params</span><span class="p">:</span>
              <span class="n">should_run</span><span class="p">:</span> <span class="p">[</span><span class="kc">True</span><span class="p">]</span>
        <span class="o">-</span> <span class="n">name</span><span class="p">:</span> <span class="s2">&quot;SCODE&quot;</span>
          <span class="n">params</span><span class="p">:</span>
              <span class="n">should_run</span><span class="p">:</span> <span class="p">[</span><span class="kc">False</span><span class="p">]</span>
              <span class="n">z</span><span class="p">:</span> <span class="p">[</span><span class="mi">10</span><span class="p">]</span>
              <span class="n">nIter</span><span class="p">:</span> <span class="p">[</span><span class="mi">1000</span><span class="p">]</span>
              <span class="n">nRep</span><span class="p">:</span> <span class="p">[</span><span class="mi">6</span><span class="p">]</span>

<span class="n">output_settings</span><span class="p">:</span>
    <span class="n">output_dir</span> <span class="p">:</span> <span class="s2">&quot;outputs&quot;</span>
    <span class="n">output_prefix</span><span class="p">:</span> <span class="s2">&quot;GSD&quot;</span>
</pre></div>
</div>
<p>BEELINE would interepret this as follows:</p>
<ul class="simple">
<li>The input expression data file is located at <code class="docutils literal notranslate"><span class="pre">inputs/example/GSD/ExpressionData.csv</span></code>, the pseudotime file is located at <code class="docutils literal notranslate"><span class="pre">inputs/example/GSD/PseudoTime.csv</span></code>, and the reference network or true edges file is located at <code class="docutils literal notranslate"><span class="pre">inputs/example/GSD/refNetwork.csv</span></code>. Note that the paths are relative to the current working directory.</li>
<li>The algorithm specific inputs will be placed under <code class="docutils literal notranslate"><span class="pre">inputs/example/GSD/&lt;algorithm_name&gt;</span></code>. For example, for PIDC, the inputs will be placed under <code class="docutils literal notranslate"><span class="pre">inputs/example/GSD/PIDC/</span></code>. The SCODE algorithm will be skipped becuase the should_run flag is set to False.</li>
<li>The output folder structure will be similar to that of the inputs under output_dir. For example, the outputs obtained after running PIDC on this dataset will be placed under <code class="docutils literal notranslate"><span class="pre">outputs/example/GSD/PIDC/</span></code>.</li>
</ul>
<div class="admonition attention">
<p class="first admonition-title">Attention</p>
<p class="last">Please ensure that the YAML file is correctly indented!</p>
</div>
</div>
<div class="section" id="running-the-pipeline">
<h3>Running the pipeline<a class="headerlink" href="#running-the-pipeline" title="Permalink to this headline">¶</a></h3>
<p>Once the input dataset has been generated and formatted as described
in Section <a class="reference internal" href="#input-datasets"><span class="std std-ref">Input Datasets</span></a> , and the config file has been
created as described in <a class="reference internal" href="#configfiles"><span class="std std-ref">Config files</span></a>, the pipeline can be
executed by simply calling <code class="docutils literal notranslate"><span class="pre">BLRun.py</span></code> with the config file
specifying the inputs and the algorithms to run, passed using the
<code class="docutils literal notranslate"><span class="pre">--config</span></code> option which takes the path to the config file.</p>
<p>To run the pipeline, simply invoke</p>
<div class="code bash highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">BLRunner</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">config</span> <span class="n">PATH</span><span class="o">/</span><span class="n">TO</span><span class="o">/</span><span class="n">CONFIG</span><span class="o">/</span><span class="n">FILE</span>
</pre></div>
</div>
<p>For details about the implementation of <a class="reference internal" href="BLRun.html#module-BLRun" title="BLRun"><code class="xref py py-class docutils literal notranslate"><span class="pre">BLRun</span></code></a> , see <span class="xref std std-ref">blrunguide</span> .</p>
</div>
<div class="section" id="running-the-evaluation-scripts">
<h3>Running the evaluation scripts<a class="headerlink" href="#running-the-evaluation-scripts" title="Permalink to this headline">¶</a></h3>
<p>Each algorithm outputs an inferred network in the form of a ranked edge list.
BEELINE implements a consistent interface using the config file in order to retrieve
the predictions of multiple algorithms and evaluate them using a variety of methods.</p>
<p>The evaluation of the inferred networks is done by calling the
<code class="docutils literal notranslate"><span class="pre">BLEvaluator.py</span></code> script.  Like the <code class="docutils literal notranslate"><span class="pre">BLRunner.py</span></code> script, the
Evaluator script takes the config file as input. Every  subsequent
option passed to this script calls a different evaluation script. For instance,
in order to analyze the AUROC and AUPRC values, and also analyze network motifs,
use the following command</p>
<div class="code bash highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">BLEvaluator</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">config</span> <span class="n">PATH</span><span class="o">/</span><span class="n">TO</span><span class="o">/</span><span class="n">CONFIG</span><span class="o">/</span><span class="n">FILE</span> \
                           <span class="o">--</span><span class="n">auc</span> \ <span class="c1"># calls the computeAUC script</span>
                           <span class="o">--</span><span class="n">motifs</span> \ <span class="c1"># calls the computeNetMotifs script</span>
</pre></div>
</div>
<p>The full list of available evaluation functions and their corresponding options to
be passed to <code class="docutils literal notranslate"><span class="pre">BLEvaluator.py</span></code> are given below:</p>
<table border="1" class="colwidths-given docutils">
<colgroup>
<col width="38%" />
<col width="63%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>-h, –help</td>
<td>show the help message and exit</td>
</tr>
<tr class="row-even"><td>-c, –config &lt;file-name&gt;</td>
<td>Configuration file containing list of datasets, algorithms, and output specifications.</td>
</tr>
<tr class="row-odd"><td>-a, –auc</td>
<td>Compute median of areas under Precision-Recall and ROC curves. Calls <a class="reference internal" href="BLEval.html#module-BLEval.computeAUC" title="BLEval.computeAUC"><code class="xref py py-mod docutils literal notranslate"><span class="pre">BLEval.computeAUC</span></code></a>.</td>
</tr>
<tr class="row-even"><td>-j, –jaccard</td>
<td>Compute median Jaccard index of predicted top-k networks
for each algorithm for a given set of datasets generated from the same
ground truth network. Calls <a class="reference internal" href="BLEval.html#module-BLEval.computeJaccard" title="BLEval.computeJaccard"><code class="xref py py-mod docutils literal notranslate"><span class="pre">BLEval.computeJaccard</span></code></a>.</td>
</tr>
<tr class="row-odd"><td>-r, –spearman</td>
<td>Compute median Spearman Corr. of predicted edges for each algorithm for a given set of datasets generated from the same ground truth network.  Calls <a class="reference internal" href="BLEval.html#module-BLEval.computeSpearman" title="BLEval.computeSpearman"><code class="xref py py-mod docutils literal notranslate"><span class="pre">BLEval.computeSpearman</span></code></a>.</td>
</tr>
<tr class="row-even"><td>-t, –time</td>
<td>Analyze time taken by each algorithm for a. Calls <a class="reference internal" href="BLEval.html#module-BLEval.parseTime" title="BLEval.parseTime"><code class="xref py py-mod docutils literal notranslate"><span class="pre">BLEval.parseTime</span></code></a>.</td>
</tr>
<tr class="row-odd"><td>-e, –epr</td>
<td>Compute median early precision. Calls <a class="reference internal" href="BLEval.html#module-BLEval.computeEarlyPrec" title="BLEval.computeEarlyPrec"><code class="xref py py-mod docutils literal notranslate"><span class="pre">BLEval.computeEarlyPrec</span></code></a>.</td>
</tr>
<tr class="row-even"><td>-s, –sepr</td>
<td>Analyze median (signed) early precision for activation and inhibitory edges. <a class="reference internal" href="BLEval.html#module-BLEval.computeSignedEPrec" title="BLEval.computeSignedEPrec"><code class="xref py py-mod docutils literal notranslate"><span class="pre">BLEval.computeSignedEPrec</span></code></a>.</td>
</tr>
<tr class="row-odd"><td>-m, –motifs</td>
<td>Compute network motifs in the predicted top-k networks. Calls <a class="reference internal" href="BLEval.html#module-BLEval.computeNetMotifs" title="BLEval.computeNetMotifs"><code class="xref py py-mod docutils literal notranslate"><span class="pre">BLEval.computeNetMotifs</span></code></a></td>
</tr>
</tbody>
</table>
<p>For details about the implementation of <a class="reference internal" href="BLEval.html#module-BLEval" title="BLEval"><code class="xref py py-class docutils literal notranslate"><span class="pre">BLEval</span></code></a> , see <a class="reference internal" href="beeline-developer.html#blevalguide"><span class="std std-ref">Adding a new evaluation technique</span></a> .</p>
</div>
</div>
<div class="section" id="details-of-supported-algorithms">
<span id="algorithms"></span><h2>Details of supported algorithms<a class="headerlink" href="#details-of-supported-algorithms" title="Permalink to this headline">¶</a></h2>
<p>The following table lists the  algorithms and the parameters they take as input, along with the default parameter values</p>
<table border="1" class="docutils">
<colgroup>
<col width="15%" />
<col width="85%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head"><strong>Algorithms</strong></th>
<th class="head"><strong>Input Parameters</strong></th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>SINCERITIES</td>
<td><ul class="first last simple">
<li><code class="docutils literal notranslate"><span class="pre">nBins</span></code> : (Default = 10)</li>
</ul>
</td>
</tr>
<tr class="row-odd"><td>SCODE</td>
<td><ul class="first last simple">
<li><code class="docutils literal notranslate"><span class="pre">z</span></code> : (Default = 10)</li>
<li><code class="docutils literal notranslate"><span class="pre">nIter</span></code> : (Default = 1000)</li>
<li><code class="docutils literal notranslate"><span class="pre">nRep</span></code> : (Default = 6)</li>
</ul>
</td>
</tr>
<tr class="row-even"><td>SCNS</td>
<td>None</td>
</tr>
<tr class="row-odd"><td>SCINGE</td>
<td><ul class="first last simple">
<li><code class="docutils literal notranslate"><span class="pre">lambda</span></code> : (Default = 0.01)</li>
<li><code class="docutils literal notranslate"><span class="pre">dT</span></code> : (Default = 15)</li>
<li><code class="docutils literal notranslate"><span class="pre">num_lags</span></code> : (Default = 5)</li>
<li><code class="docutils literal notranslate"><span class="pre">kernel_width</span></code> : (Default = 0.5)</li>
<li><code class="docutils literal notranslate"><span class="pre">prob_zero_removal</span></code> : (Default = 0)</li>
<li><code class="docutils literal notranslate"><span class="pre">prob_remove_samples</span></code> : (Default = 0.0)</li>
<li><code class="docutils literal notranslate"><span class="pre">family</span></code> : (Default = ‘gaussian’)</li>
<li><code class="docutils literal notranslate"><span class="pre">num_replicates</span></code> : (Default = 6)</li>
</ul>
</td>
</tr>
<tr class="row-even"><td>PPCOR</td>
<td><ul class="first last simple">
<li><code class="docutils literal notranslate"><span class="pre">pVal</span></code> : p-value cutoff (Default = 0.01)</li>
</ul>
</td>
</tr>
<tr class="row-odd"><td>PIDC</td>
<td>None</td>
</tr>
<tr class="row-even"><td>LEAP</td>
<td><ul class="first last simple">
<li><code class="docutils literal notranslate"><span class="pre">maxLag</span></code> : (Default = 0.33)</li>
</ul>
</td>
</tr>
<tr class="row-odd"><td>SCRIBE</td>
<td><ul class="first last simple">
<li><code class="docutils literal notranslate"><span class="pre">delay</span></code> :  (Default = 5 )</li>
<li><code class="docutils literal notranslate"><span class="pre">method</span></code> : Any of ‘RDI’, ‘uRDI’, ‘cRDI’, or ‘ucRDI’ (Default = ‘ucRDI’)</li>
<li><code class="docutils literal notranslate"><span class="pre">lowerDetectionLimit</span></code> : (Default = 0)</li>
<li><code class="docutils literal notranslate"><span class="pre">expressionFamily</span></code> : If mRNA counts, use ‘negbinomial’ (Default = ‘uninormal’)</li>
<li><code class="docutils literal notranslate"><span class="pre">log</span></code> : Log transform expression values (Default = False)</li>
<li><code class="docutils literal notranslate"><span class="pre">ignorePT</span></code> : Ignore Pseudotime (Default = True)</li>
</ul>
</td>
</tr>
<tr class="row-even"><td>GRNVBEM</td>
<td>None</td>
</tr>
<tr class="row-odd"><td>GRISLI</td>
<td><ul class="first last simple">
<li><code class="docutils literal notranslate"><span class="pre">L</span></code> : (Default = 10)</li>
<li><code class="docutils literal notranslate"><span class="pre">R</span></code> : (Default = 3000)</li>
<li><code class="docutils literal notranslate"><span class="pre">alphaMin</span></code> : (Default = 0.0)</li>
</ul>
</td>
</tr>
<tr class="row-even"><td>GENIE3</td>
<td>None</td>
</tr>
<tr class="row-odd"><td>GRNBOOST2</td>
<td>None</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="developer-guide">
<h2>Developer Guide<a class="headerlink" href="#developer-guide" title="Permalink to this headline">¶</a></h2>
<p id="blrunguide">Please follow the following steps to add a new GRN inferenmence method to BEELINE.</p>
<div class="section" id="adding-a-new-grn-inference-method">
<h3>Adding a new GRN inference method<a class="headerlink" href="#adding-a-new-grn-inference-method" title="Permalink to this headline">¶</a></h3>
<p>In order to ensure easy set-up and portability, all the GRN algorithms in BEELINE are provided within their own separate Docker instances. This also avoids conflicting libraries/software versions that may arise from the GRN algorithm implmentations. More resources on Docker can be found <a class="reference external" href="https://docs.docker.com/get-started/resources/">here</a>. In order to add a new GRN method to BEELINE, you will need the following components.</p>
<ol class="arabic simple">
<li><strong>Create a Dockerfile:</strong> To create a Docker image for your GRN algorithm locally, you will need to start with a “Dockerfile” that contains necessary software specifications and commands listed in a specific order from top to bottom. <a class="reference external" href="https://github.com/Murali-group/Beeline/tree/master/Algorithms/">BEELINE</a> currently provides Dockerfiles for GRN implementations in <a class="reference external" href="https://github.com/Murali-group/Beeline/tree/master/Algorithms/ARBORETO">Python</a>, <a class="reference external" href="https://github.com/Murali-group/Beeline/tree/master/Algorithms/PPCOR">R</a>, <a class="reference external" href="https://github.com/Murali-group/Beeline/blob/master/Algorithms/GRISLI/Dockerfile">MATLAB</a>, and <a class="reference external" href="https://github.com/Murali-group/Beeline/tree/master/Algorithms/PIDC">Julia</a>. These Dockerfiles can be used as a template for adding other GRN algorithms. Note that in order to run MATLAB-based code, we recommend first builing a standalone MATLAB application for your GRN algorithm  <a class="reference external" href="https://www.mathworks.com/help/compiler/create-and-install-a-standalone-application-from-matlab-code.html">MATLAB Runtime</a> and then setting up your Dockerfile to simply execute the pre-compiled binaries. For example, the <a class="reference external" href="https://github.com/Murali-group/Beeline/blob/master/Algorithms/PPCOR/Dockerfile">Dockerfile</a> using R which runs the script <a class="reference external" href="https://github.com/Murali-group/Beeline/blob/master/Algorithms/PPCOR/runPPCOR.R">runPPCOR.R</a> within the Docker container is as follows:</li>
</ol>
<div class="code Dockerfile highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">FROM</span> <span class="n">r</span><span class="o">-</span><span class="n">base</span><span class="p">:</span><span class="mf">3.5</span><span class="o">.</span><span class="mi">3</span> <span class="c1">#This is the base image upon which necessary libraries are installed</span>

<span class="n">LABEL</span> <span class="n">maintainer</span> <span class="o">=</span> <span class="s2">&quot;Aditya Pratapa &lt;adyprat@vt.edu&gt;&quot;</span> <span class="c1">#Additional information</span>

<span class="n">RUN</span> <span class="n">apt</span><span class="o">-</span><span class="n">get</span> <span class="n">update</span> <span class="o">&amp;&amp;</span> <span class="n">apt</span><span class="o">-</span><span class="n">get</span> <span class="n">install</span> <span class="n">time</span> <span class="c1">#Installl time command to compute time taken</span>

<span class="n">USER</span> <span class="n">root</span> <span class="c1">#Set main user as root to avoid permission issues</span>

<span class="n">WORKDIR</span> <span class="o">/</span> <span class="c1">#Sets current working directory</span>

<span class="n">RUN</span> <span class="n">R</span> <span class="o">-</span><span class="n">e</span> <span class="s2">&quot;install.packages(&#39;https://cran.r-project.org/src/contrib/ppcor_1.1.tar.gz&#39;, type = &#39;source&#39;)&quot;</span> <span class="c1">#Installs a specific version of PPCOR package</span>

<span class="n">COPY</span> <span class="n">runPPCOR</span><span class="o">.</span><span class="n">R</span> <span class="o">/</span> <span class="c1">#Copy the main R script which will perform necessary GRN computations</span>

<span class="n">RUN</span> <span class="n">mkdir</span> <span class="n">data</span><span class="o">/</span> <span class="c1">#Make a directory to mount the folder containing input files</span>
</pre></div>
</div>
<p>The Dockerfile must then be placed under the <a class="reference external" href="https://github.com/Murali-group/Beeline/tree/master/Algorithms/">Algorithms/&lt;algorithm-name&gt;</a> folder in BEELINE. Then the following lines must be added to the <a class="reference external" href="https://github.com/Murali-group/Beeline/blob/master/initialize.sh">initialize.sh</a> script.</p>
<div class="code bash highlight-default notranslate"><div class="highlight"><pre><span></span>cd $BASEDIR/Algorithms/&lt;algorithm-name&gt;/
docker build -q -t &lt;algorithm-name&gt;:base .
echo &quot;Docker container for &lt;algorithm-name&gt; is built and tagged as &lt;algorithm-name&gt;:base&quot;
</pre></div>
</div>
<p>Replace &lt;algorithm-name&gt; with the name of the new algorithm that is being added to BEELINE.</p>
<p>2. <strong>Create a &lt;algorithm-name&gt;Runner.py script:</strong> Once the Docker image is built locally using the above Dockerfile, we will then need to setup a <a class="reference internal" href="BLRun.html#module-BLRun" title="BLRun"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BLRun</span></code></a> object which will read necessary inputs, runs GRN algorithm inside the Docker image, and finally parses output so that evaluation can be performed using <a class="reference internal" href="BLEval.html#module-BLEval" title="BLEval"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BLEval</span></code></a>.
Most of the GRN algorithms in BEELINE require a gene-by-cell matrix provided as input, along with a pesudotime ordering of cells, and any additional manually specified parameters. These details are specified as inputs to BEELINE using <a class="reference internal" href="#configfiles"><span class="std std-ref">Config files</span></a> as mentioned earlier. In our current implementation, we provide a separate csv file for expression matrix and pseudotime files, whereas algorithm parameters are specified as command line arguments. However, this can be modified easily to specify even the parameters using a separate file by simply providing its path in the config file. Each &lt;algorithm-name&gt;Runner.py script should contain the following three functions:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">generateInputs()</span></code> : This function reads the two input data files (i.e., expression data and the pseudotime), and processes them into the format required by the given algorithm. For example, the algorithm may only require the cells in the expression matrix to be ordered in pseudotime, instead of exact pesudotime values as input. Moreover, if the algorithm requires that you run the GRN inference on each trajectory separately, we can this function to write the separate expression matrices contianing only cells  of a particular trajectory.</li>
<li><code class="docutils literal notranslate"><span class="pre">run()</span></code> : This function constructs a “docker run” system command with the appropriate command line parameters to be passed to the docker container in order to run a given algorithm. The docker run command also needs to mount the folder containing the input files using the “-v” flag and specify where the outputs are written. If the algorithm needs to be run separately on each trajectory, it can be simply called in a loop inside this function.</li>
<li><code class="docutils literal notranslate"><span class="pre">parseOutput()</span></code> : This function reads the algorithm-specific outputs and formats it into a ranked edgelist comma-separated file in the following format which can be subsequently used by <a class="reference internal" href="BLEval.html#module-BLEval" title="BLEval"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BLEval</span></code></a></li>
</ul>
</div></blockquote>
<div class="code text highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Gene1</span><span class="p">,</span><span class="n">Gene2</span><span class="p">,</span><span class="n">EdgeWeight</span>
<span class="n">reg1</span><span class="p">,</span><span class="n">targ1</span><span class="p">,</span><span class="n">edgeweight</span>
</pre></div>
</div>
<p>where the first line are the column names, and the subsequent lines contain the edges predicted by the network. The Gene1 column should contain regulators, the Gene2 column the targets, and EdgeWeight column the absolute value of the weight predicted for edge (regulator,target). Note that in cases where the algorithm requires that you run the GRN inference on each trajectory separately, a single output network is obtained by finding the maximum score for a given edge across the GRNs computed for each individual trajectory.  Please ensure that all of the above three functions in your script accept arguments of type <a class="reference internal" href="BLRun.html#BLRun.runner.Runner" title="BLRun.runner.Runner"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BLRun.runner.Runner</span></code></a>. The &lt;algorithm-name&gt;Runner.py script must then be placed under the <a class="reference external" href="https://github.com/Murali-group/Beeline/tree/master/BLRun/">BLRun/</a>  folder in BEELINE.</p>
<ol class="arabic simple" start="3">
<li><strong>Add the new alorithm to runner.py:</strong> The next step is to integrate the new algorithm within <a class="reference internal" href="BLRun.html#module-BLRun" title="BLRun"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BLRun</span></code></a> object. This can be achieved by adding the above three modules from the above step, i.e, <code class="docutils literal notranslate"><span class="pre">generateInputs()</span></code>, <code class="docutils literal notranslate"><span class="pre">run()</span></code>, and <code class="docutils literal notranslate"><span class="pre">parseOutput()</span></code> to <a class="reference external" href="https://github.com/Murali-group/Beeline/blob/master/BLRun/runner.py">runner.py</a>.</li>
<li><strong>Add the new alorithm to config.yaml:</strong> The final step is to add the new algorithm and any necessary parameters to the cofig.yaml. Note that currently BEELINE can only handle one parameter set at a time eventhough multiple parameters can be passed onto the single parameter object.</li>
</ol>
<div class="code text highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">algorithms</span><span class="p">:</span>
    <span class="o">-</span> <span class="n">name</span><span class="p">:</span> <span class="o">&lt;</span><span class="n">algorithm</span><span class="o">-</span><span class="n">name</span><span class="o">&gt;</span>
      <span class="n">params</span><span class="p">:</span>
          <span class="c1"># Any parameters that can be passed to this algorithm</span>
          <span class="n">paramX</span><span class="p">:</span> <span class="p">[</span><span class="mf">0.1</span><span class="p">]</span>
          <span class="n">paramY</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;Text&#39;</span><span class="p">]</span>
          <span class="n">should_run</span><span class="p">:</span> <span class="p">[</span><span class="kc">True</span><span class="p">]</span> <span class="c1"># or False</span>
</pre></div>
</div>
</div>
<div class="section" id="adding-a-new-evaluation-technique">
<span id="blevalguide"></span><h3>Adding a new evaluation technique<a class="headerlink" href="#adding-a-new-evaluation-technique" title="Permalink to this headline">¶</a></h3>
<p>BEELINE currently supports several evaluation techniques, namely, area under ROC and PR curves, eary precision, early signed precision, time and memory consumption, network motifs, jaccard similarities, and spearman coefficients. A new evaluation technique can be easily added to BEELINE using the following procedure.</p>
<ol class="arabic simple">
<li>Add the script containing the new evaluation technique to the <a class="reference external" href="https://github.com/Murali-group/Beeline/tree/master/BLEval/">BLEval/</a> folder.</li>
<li>The next step is to integrate the new technique into the <a class="reference internal" href="BLEval.html#module-BLEval" title="BLEval"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BLEval</span></code></a> object. This can be achieved by adding it as a new module under <a class="reference internal" href="BLEval.html#module-BLEval" title="BLEval"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BLEval</span></code></a> in the <a class="reference external" href="https://github.com/Murali-group/Beeline/blob/master/BLEval/__init__.py">BLEval/__init__.py</a> script. Please ensure that your script can accept arguments of type <a class="reference internal" href="BLEval.html#module-BLEval" title="BLEval"><code class="xref py py-obj docutils literal notranslate"><span class="pre">BLEval</span></code></a>.</li>
<li>The final step is to add a command line option to perform the evaluation to <a class="reference external" href="https://github.com/Murali-group/Beeline/blob/master/BLEvaluator.py">BLEvaluator.py</a>.</li>
</ol>
</div>
</div>
<div class="section" id="api-reference">
<h2>API Reference<a class="headerlink" href="#api-reference" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="BLRun.html">BLRun package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="BLRun.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLRun.html#module-BLRun.genie3Runner">BLRun.genie3Runner module</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLRun.html#module-BLRun.grisliRunner">BLRun.grisliRunner module</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLRun.html#module-BLRun.grnboost2Runner">BLRun.grnboost2Runner module</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLRun.html#module-BLRun.grnvbemRunner">BLRun.grnvbemRunner module</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLRun.html#module-BLRun.jump3Runner">BLRun.jump3Runner module</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLRun.html#module-BLRun.leapRunner">BLRun.leapRunner module</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLRun.html#module-BLRun.pidcRunner">BLRun.pidcRunner module</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLRun.html#module-BLRun.ppcorRunner">BLRun.ppcorRunner module</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLRun.html#module-BLRun.runner">BLRun.runner module</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLRun.html#module-BLRun.scingeRunner">BLRun.scingeRunner module</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLRun.html#module-BLRun.scnsRunner">BLRun.scnsRunner module</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLRun.html#module-BLRun.scodeRunner">BLRun.scodeRunner module</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLRun.html#module-BLRun.scribeRunner">BLRun.scribeRunner module</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLRun.html#module-BLRun.sinceritiesRunner">BLRun.sinceritiesRunner module</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BLEval.html">BLEval package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="BLEval.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLEval.html#module-BLEval.computeAUC">BLEval.computeAUC module</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLEval.html#module-BLEval.computeDGAUC">BLEval.computeDGAUC module</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLEval.html#module-BLEval.computeEarlyPrec">BLEval.computeEarlyPrec module</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLEval.html#module-BLEval.computeJaccard">BLEval.computeJaccard module</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLEval.html#module-BLEval.computeNetMotifs">BLEval.computeNetMotifs module</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLEval.html#module-BLEval.computeSignedEPrec">BLEval.computeSignedEPrec module</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLEval.html#module-BLEval.computeSpearman">BLEval.computeSpearman module</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLEval.html#module-BLEval.parseTime">BLEval.parseTime module</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BLPlot.html">BLPlot package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="BLPlot.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLPlot.html#module-BLPlot.CuratedOverview">BLPlot.CuratedOverview module</a></li>
<li class="toctree-l2"><a class="reference internal" href="BLPlot.html#module-BLPlot.NetworkMotifs">BLPlot.NetworkMotifs module</a></li>
</ul>
</li>
</ul>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="BLRun.html" class="btn btn-neutral float-right" title="BLRun package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="index.html" class="btn btn-neutral float-left" title="Welcome to BEELINE’s documentation!" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Aditya Pratapa

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>