

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>BoolODE package &mdash; BEELINE  documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="BoolODE" href="BoolODE-intro.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> BEELINE
          

          
            
            <img src="_static/beeline.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="BEELINE.html">BEELINE</a></li>
<li class="toctree-l1"><a class="reference internal" href="BoolODE-intro.html">BoolODE</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">BEELINE</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="BoolODE-intro.html">BoolODE</a> &raquo;</li>
        
      <li>BoolODE package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/BoolODE.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-BoolODE">
<span id="boolode-package"></span><span id="boolode"></span><h1>BoolODE package<a class="headerlink" href="#module-BoolODE" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="BoolODE.BoolODE">
<em class="property">class </em><code class="descclassname">BoolODE.</code><code class="descname">BoolODE</code><span class="sig-paren">(</span><em>job_settings: BoolODE.JobSettings</em>, <em>global_settings: BoolODE.GlobalSettings</em>, <em>postproc_settings: BoolODE.PostProcSettings</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE.html#BoolODE"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.BoolODE" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>The BoolODE object is created by parsing a user-provided configuration
file. It then calls the correct functions for executing the user defined
experiment.</p>
<dl class="method">
<dt id="BoolODE.BoolODE.do_post_processing">
<code class="descname">do_post_processing</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE.html#BoolODE.do_post_processing"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.BoolODE.do_post_processing" title="Permalink to this definition">¶</a></dt>
<dd><p>Call genSamples() first. Then run DimRed runSlingShot,  
generateDropouts, if specified by the user.</p>
</dd></dl>

<dl class="method">
<dt id="BoolODE.BoolODE.execute_jobs">
<code class="descname">execute_jobs</code><span class="sig-paren">(</span><em>parallel=False</em>, <em>num_threads=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE.html#BoolODE.execute_jobs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.BoolODE.execute_jobs" title="Permalink to this definition">¶</a></dt>
<dd><p>Run each user specified job. 
BoolODE runs two types of functions
1. If <cite>do_simulation == TRUE</cite>, perform SDE simulations of model specified as Boolean rules. 
2. If <cite>do_post_processing == TRUE</cite> perform the list of post processing operations specified.</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">This function automatically creates folders for each job name 
as specified in the config file, if the folder doesn’t already exist.
Contents of existing folders will be rewritten!</p>
</div>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="BoolODE.ConfigParser">
<em class="property">class </em><code class="descclassname">BoolODE.</code><code class="descname">ConfigParser</code><a class="reference internal" href="_modules/BoolODE.html#ConfigParser"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.ConfigParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Define static methods for parsing a config file that sets a large number
of parameters for the pipeline</p>
<dl class="staticmethod">
<dt id="BoolODE.ConfigParser.parse">
<em class="property">static </em><code class="descname">parse</code><span class="sig-paren">(</span><em>config_file_handle</em><span class="sig-paren">)</span> &#x2192; BoolODE.BoolODE<a class="reference internal" href="_modules/BoolODE.html#ConfigParser.parse"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.ConfigParser.parse" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="BoolODE.GlobalSettings">
<em class="property">class </em><code class="descclassname">BoolODE.</code><code class="descname">GlobalSettings</code><span class="sig-paren">(</span><em>model_dir</em>, <em>output_dir</em>, <em>do_simulations</em>, <em>do_post_processing</em>, <em>modeltype</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE.html#GlobalSettings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.GlobalSettings" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
</dd></dl>

<dl class="class">
<dt id="BoolODE.JobSettings">
<em class="property">class </em><code class="descclassname">BoolODE.</code><code class="descname">JobSettings</code><span class="sig-paren">(</span><em>jobs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE.html#JobSettings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.JobSettings" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Store user defined job settings</p>
</dd></dl>

<dl class="class">
<dt id="BoolODE.PostProcSettings">
<em class="property">class </em><code class="descclassname">BoolODE.</code><code class="descname">PostProcSettings</code><span class="sig-paren">(</span><em>dropout_jobs</em>, <em>dimred_jobs</em>, <em>slingshot_jobs</em>, <em>gensample_jobs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE.html#PostProcSettings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.PostProcSettings" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
</dd></dl>

<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-BoolODE.model_generator">
<span id="boolode-model-generator-module"></span><h2>BoolODE.model_generator module<a class="headerlink" href="#module-BoolODE.model_generator" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="BoolODE.model_generator.GenerateModel">
<em class="property">class </em><code class="descclassname">BoolODE.model_generator.</code><code class="descname">GenerateModel</code><span class="sig-paren">(</span><em>settings</em>, <em>parameterInputsDF</em>, <em>parameterSetDF</em>, <em>interactionStrengthDF</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/model_generator.html#GenerateModel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.model_generator.GenerateModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.8)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Class that holds model attributes. Provides helper functions to convert a Boolean model
to an ODE model</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>settings</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.8)"><em>dict</em></a>) – User defined job settings. This is created in boolode.do_post_processing()</li>
<li><strong>parameterInputsDF</strong> (<em>pandas DataFrame</em>) – User specified parmeter inputs. Specified in config YAML as a path.</li>
<li><strong>parameterSetDF</strong> (<em>pandas DataFrame</em>) – User specified kinetic parameters. Specified in config YAML as a path.</li>
<li><strong>interactionStrengthDF</strong> (<em>pandas DataFrame</em>) – User specified interaction strengths. Specified in config YAML as a path</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>This class is initialized with the job_settings dictionary created 
by the BoolODE class. Additionally, three dataframe objects are required:</p>
<ol class="arabic simple">
<li>parameterInputsDF - specifies parameter inputs</li>
<li>parameterSetDF - specifies pregenerated python file containing kinetic parameter values</li>
<li>interactionStrengthDF - specifies interactions strengths</li>
</ol>
<p>By default, if these aren’t specified by the user, BoolODE stores empty pandas DataFrame objects.</p>
<dl class="method">
<dt id="BoolODE.model_generator.GenerateModel.addDummyGenes">
<code class="descname">addDummyGenes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/model_generator.html#GenerateModel.addDummyGenes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.model_generator.GenerateModel.addDummyGenes" title="Permalink to this definition">¶</a></dt>
<dd><p>Add dummy genes. 
Starts with the user specified Boolean model, and grows the
network by adding new ‘dummy’ nodes to each existing node.
In the current configuration, the user can specify <cite>max_parents</cite>
which specifies the number of parent nodes of every dummy node.
Currently, twice as many dummy nodes as nodes in the original graph
are added.</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">This feature is experimental.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="BoolODE.model_generator.GenerateModel.addInteractionStrengths">
<code class="descname">addInteractionStrengths</code><span class="sig-paren">(</span><em>hillThreshold</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/model_generator.html#GenerateModel.addInteractionStrengths"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.model_generator.GenerateModel.addInteractionStrengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if interaction strengths are specified.
Create a new parameter specifying the strength
of a given interaction. For all other equations in
which “regulator” appears in, its hillThreshold value
will remain the default value.</p>
</dd></dl>

<dl class="method">
<dt id="BoolODE.model_generator.GenerateModel.addParameterInputs">
<code class="descname">addParameterInputs</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/model_generator.html#GenerateModel.addParameterInputs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.model_generator.GenerateModel.addParameterInputs" title="Permalink to this definition">¶</a></dt>
<dd><p>Check for user specified input parameters.
If a parameter input file is specified,
Every row in parameterInputsDF contains two
columns:</p>
<ol class="arabic simple">
<li>‘Input’ specifies the name of input parameter</li>
<li>‘Value’ specified the input state, typically 1 or 0, but for HIGH and LOW inputs respectively, but any float between 0 and 1 is valid input.</li>
</ol>
<p>Now, we create a fake hill term for this input, even
though there is no equation corresponding to it. Thus,
we create a hillThreshold and hillCoefficient 
term for it. This is useful so we can treat this input
parameter as a “regulator” and leave the structure of
the logic term unchanged.</p>
</dd></dl>

<dl class="method">
<dt id="BoolODE.model_generator.GenerateModel.assignDefaultParameterValues">
<code class="descname">assignDefaultParameterValues</code><span class="sig-paren">(</span><em>parameterNamePrefixAndDefaultsAll</em>, <em>parameterNamePrefixAndDefaultsGenes</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/model_generator.html#GenerateModel.assignDefaultParameterValues"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.model_generator.GenerateModel.assignDefaultParameterValues" title="Permalink to this definition">¶</a></dt>
<dd><p>Set each kinetic parameter to its default value.</p>
</dd></dl>

<dl class="method">
<dt id="BoolODE.model_generator.GenerateModel.assignSampledParameterValues">
<code class="descname">assignSampledParameterValues</code><span class="sig-paren">(</span><em>parameterNamePrefixAndDefaultsAll</em>, <em>parameterNamePrefixAndDefaultsGenes</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/model_generator.html#GenerateModel.assignSampledParameterValues"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.model_generator.GenerateModel.assignSampledParameterValues" title="Permalink to this definition">¶</a></dt>
<dd><p>Sample kinetic parameters from a truncated normal distribution with mean=default
parameter value, and standard deviation = sample_std, in a range +- 10% of 
the default value.</p>
</dd></dl>

<dl class="method">
<dt id="BoolODE.model_generator.GenerateModel.createRegulatoryTerms">
<code class="descname">createRegulatoryTerms</code><span class="sig-paren">(</span><em>currgene</em>, <em>combinationOfRegulators</em>, <em>regSpecies</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/model_generator.html#GenerateModel.createRegulatoryTerms"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.model_generator.GenerateModel.createRegulatoryTerms" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates the terms in the activation function which governs
the transcriptional state of the gene. 
If modeltype is ‘hill’, each term is composed of products of Hill functions, ranging over
all possible combinations of regulators. 
Similarly, if modeltype is ‘heaviside’, each term is simply a
product of the activities of the combinations of regulators, for all combinations.
Here, the form of a single term is generated for a single combination of regulators.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>currgene</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – Name of the current gene</li>
<li><strong>combinationOfRegulators</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a>) – a list of all combinations of regulators of currgene</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="BoolODE.model_generator.GenerateModel.generateModelDict">
<code class="descname">generateModelDict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/model_generator.html#GenerateModel.generateModelDict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.model_generator.GenerateModel.generateModelDict" title="Permalink to this definition">¶</a></dt>
<dd><p>Take a DataFrame object with Boolean rules,
construct ODE equations for each variable.
This is the core function in BoolODE.
In a nutshell, a temporary namespace called boolodespace
is created. In this namespace, the nodes  in the Boolean network
are first initialized to 0, or OFF. Then, for each gene/node,
its corresponding rule is evaluated by setting its regulators to
all combinations binary states. The outcome of each of these Boolean
rule evaluations is used to decide the value of the activation strength
parameter ‘a’ in Hill functions, or the interaction parameter ‘w’ in 
the Heaviside functions. 
Each step of this conversion is documented in the code directly.</p>
</dd></dl>

<dl class="method">
<dt id="BoolODE.model_generator.GenerateModel.getParameters">
<code class="descname">getParameters</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/model_generator.html#GenerateModel.getParameters"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.model_generator.GenerateModel.getParameters" title="Permalink to this definition">¶</a></dt>
<dd><p>Create and assigns the kinetic parameters for each equations the
corresponding parameter value.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Default kinetic parameter values are stored in ./parameters.yaml. This
is provided to make it easy for users to define different defaults, but it 
other parameter values have not been tested extensively, and might not
yield the expected results.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="BoolODE.model_generator.GenerateModel.readBooleanRules">
<code class="descname">readBooleanRules</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/model_generator.html#GenerateModel.readBooleanRules"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.model_generator.GenerateModel.readBooleanRules" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads a rule file from path and stores the rules in a dictionary.
Performs the following transformations:</p>
<ol class="arabic simple">
<li>Parses each Boolean rule to extract nodes in the model</li>
<li>Identifies nodes without corresponding rules: (a) If a parameterInput file is specified and the node under consideration
is present in the file, the node is treated as an input parameter
(b) Else, a self-edge is added to the node</li>
<li>Adds dummy genes - <em>This is experimental</em></li>
<li>Finally, create a gene (x_) and protein (p_)  variable corresponding to 
each node in the network, taking into consideration user defined types.</li>
</ol>
</dd></dl>

<dl class="method">
<dt id="BoolODE.model_generator.GenerateModel.writeModelToFile">
<code class="descname">writeModelToFile</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/model_generator.html#GenerateModel.writeModelToFile"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.model_generator.GenerateModel.writeModelToFile" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes model to file as a python function.
The ODE model generated using generateModelDict() is defined as 
a python ODE function called Model(). Model() takes 3 arguments:</p>
<ol class="arabic simple">
<li>The current model state vector Y</li>
<li>The current time t</li>
<li>A list of parameters pars</li>
</ol>
<p>The function returns a vector of time derivatives computed from the ODEs.
Model() is written model.py in the directory of the current job</p>
</dd></dl>

<dl class="method">
<dt id="BoolODE.model_generator.GenerateModel.writeParametersToFile">
<code class="descname">writeParametersToFile</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/model_generator.html#GenerateModel.writeParametersToFile"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.model_generator.GenerateModel.writeParametersToFile" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes dictionary of parameters to file.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-BoolODE.post_processing">
<span id="boolode-post-processing-module"></span><h2>BoolODE.post_processing module<a class="headerlink" href="#module-BoolODE.post_processing" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="BoolODE.post_processing.computeSSPT">
<code class="descclassname">BoolODE.post_processing.</code><code class="descname">computeSSPT</code><span class="sig-paren">(</span><em>opts</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/post_processing.html#computeSSPT"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.post_processing.computeSSPT" title="Permalink to this definition">¶</a></dt>
<dd><p><em>Author: Aditya Pratapa</em></p>
<p>Compute PseudoTime using ‘slingshot’.
Needs the input Gene x Cells expression data frame.
Needs number of clusters to be expected in the 
dataset. 
E.g., Linear: k=1 (no PseudoTime inference is done)
E.g., Bifurcating: k=3 (1 initial and 2 terminal)
E.g., Trifurcating: k=4 (1 initial and 3 terminal)</p>
</dd></dl>

<dl class="function">
<dt id="BoolODE.post_processing.doDimRed">
<code class="descclassname">BoolODE.post_processing.</code><code class="descname">doDimRed</code><span class="sig-paren">(</span><em>opts</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/post_processing.html#doDimRed"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.post_processing.doDimRed" title="Permalink to this definition">¶</a></dt>
<dd><p>Carry out dimensionality reduction</p>
</dd></dl>

<dl class="function">
<dt id="BoolODE.post_processing.genDropouts">
<code class="descclassname">BoolODE.post_processing.</code><code class="descname">genDropouts</code><span class="sig-paren">(</span><em>opts</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/post_processing.html#genDropouts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.post_processing.genDropouts" title="Permalink to this definition">¶</a></dt>
<dd><p>Induce drop out in the simulated scRNAseq datasets.</p>
</dd></dl>

<dl class="function">
<dt id="BoolODE.post_processing.genSamples">
<code class="descclassname">BoolODE.post_processing.</code><code class="descname">genSamples</code><span class="sig-paren">(</span><em>opts</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/post_processing.html#genSamples"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.post_processing.genSamples" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate samples of cells from a given set of simulations. This sample
will then be used for other post processing steps.</p>
</dd></dl>

</div>
<div class="section" id="module-BoolODE.run_experiment">
<span id="boolode-run-experiment-module"></span><h2>BoolODE.run_experiment module<a class="headerlink" href="#module-BoolODE.run_experiment" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="BoolODE.run_experiment.Experiment">
<code class="descclassname">BoolODE.run_experiment.</code><code class="descname">Experiment</code><span class="sig-paren">(</span><em>mg</em>, <em>Model</em>, <em>tspan</em>, <em>settings</em>, <em>icsDF</em>, <em>writeProtein=False</em>, <em>normalizeTrajectory=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/run_experiment.html#Experiment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.run_experiment.Experiment" title="Permalink to this definition">¶</a></dt>
<dd><p>Carry out an <cite>in-silico</cite> experiment. This function takes as input 
an ODE model defined as a python function and carries out stochastic
simulations. BoolODE defines a _cell_ as a single time point from 
a simulated time course. Thus, in order to obtain 50 single cells,
BoolODE carries out 50 simulations, which are stored in ./simulations/.
Further, if it is known that the resulting single cell dataset will
exhibit multiple trajectories, the user can specify  the number of clusters in
<cite>nClusters</cite>; BoolODE will then cluster the entire simulation, such that each
simulated trajectory possesess a cluster ID.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>mg</strong> (<em>BoolODE.GenerateModel</em>) – Model details obtained by instantiating an object of GenerateModel</li>
<li><strong>Model</strong> (<em>function</em>) – Function defining ODE model</li>
<li><strong>tspan</strong> (<em>ndarray</em>) – Array of time points</li>
<li><strong>settings</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.8)"><em>dict</em></a>) – The job settings dictionary</li>
<li><strong>icsDF</strong> (<em>pandas DataFrame</em>) – Dataframe specifying initial condition for simulation</li>
<li><strong>writeProtein</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a>) – Bool specifying if the protein values should be written to file. Default = False</li>
<li><strong>normalizeTrajectory</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a>) – Bool specifying if the gene expression values should be scaled between 0 and 1.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="BoolODE.run_experiment.simulateAndSample">
<code class="descclassname">BoolODE.run_experiment.</code><code class="descname">simulateAndSample</code><span class="sig-paren">(</span><em>argdict</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/run_experiment.html#simulateAndSample"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.run_experiment.simulateAndSample" title="Permalink to this definition">¶</a></dt>
<dd><p>Handles parallelization of ODE simulations.
Calls the simulator with simulation settings.</p>
</dd></dl>

<dl class="function">
<dt id="BoolODE.run_experiment.startRun">
<code class="descclassname">BoolODE.run_experiment.</code><code class="descname">startRun</code><span class="sig-paren">(</span><em>settings</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/run_experiment.html#startRun"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.run_experiment.startRun" title="Permalink to this definition">¶</a></dt>
<dd><p>Start a simulation run. Loads model file, starts an Experiment(),
and generates the appropriate input files</p>
</dd></dl>

</div>
<div class="section" id="module-BoolODE.simulator">
<span id="boolode-simulator-module"></span><h2>BoolODE.simulator module<a class="headerlink" href="#module-BoolODE.simulator" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="BoolODE.simulator.deltaW">
<code class="descclassname">BoolODE.simulator.</code><code class="descname">deltaW</code><span class="sig-paren">(</span><em>N</em>, <em>m</em>, <em>h</em>, <em>seed=0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/simulator.html#deltaW"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.simulator.deltaW" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate sequence of Wiener increments for m independent Wiener
processes W_j(t) j=0..m-1 for each of N time intervals of length h.    
From the sdeint implementation</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><ul class="simple">
<li>dW : The [n, j] element has the value W_j((n+1)*h) - W_j(n*h) ( has shape (N, m) )</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="BoolODE.simulator.eulersde">
<code class="descclassname">BoolODE.simulator.</code><code class="descname">eulersde</code><span class="sig-paren">(</span><em>f</em>, <em>G</em>, <em>y0</em>, <em>tspan</em>, <em>pars</em>, <em>seed=0.0</em>, <em>dW=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/simulator.html#eulersde"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.simulator.eulersde" title="Permalink to this definition">¶</a></dt>
<dd><p>Adapted from sdeint implementation <a class="reference external" href="https://github.com/mattja/sdeint/">https://github.com/mattja/sdeint/</a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>f</strong> (<em>function</em>) – function defining ODE model. Should take vector of current state, current time, and list of parameter values as arguments.</li>
<li><strong>pars</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a>) – List of parameter values</li>
<li><strong>y0</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a>) – list of initial values</li>
<li><strong>tspan</strong> (<em>ndarray</em>) – Array of timepoints to simulate</li>
<li><strong>seed</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.8)"><em>float</em></a>) – Seed to initialize random number generator</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><ul class="simple">
<li>y: Array containing the time course of state variables</li>
</ul>
</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="BoolODE.simulator.getInitialCondition">
<code class="descclassname">BoolODE.simulator.</code><code class="descname">getInitialCondition</code><span class="sig-paren">(</span><em>ss</em>, <em>ModelSpec</em>, <em>rnaIndex</em>, <em>proteinIndex</em>, <em>genelist</em>, <em>proteinlist</em>, <em>varmapper</em>, <em>revvarmapper</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/simulator.html#getInitialCondition"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.simulator.getInitialCondition" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the initial values of all state variables. 
Takes into consideration user defined initial conditions, and computes the steady 
states of the protein variables based on the estimated values of their corresponding genes.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>ss</strong> (<em>ndarray</em>) – Steady state array</li>
<li><strong>ModelSpec</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.8)"><em>dict</em></a>) – Dictionary of dictionary specifying the ODE model, containing parameters, initial conditions and equations.</li>
<li><strong>rnaIndex</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a>) – list of indices of genes</li>
<li><strong>proteinIndex</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a>) – List of indices of proteins</li>
<li><strong>genelist</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a>) – List of names of all genes in the model</li>
<li><strong>proteinlist</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a>) – List of names of all proteins in the model</li>
<li><strong>varmapper</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.8)"><em>dict</em></a>) – Mapper: {variable name : index}</li>
<li><strong>revvarmapper</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.8)"><em>dict</em></a>) – Mapper: {index : variable name}</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><ul class="simple">
<li>newics: List containing new initial conditions</li>
</ul>
</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="BoolODE.simulator.noise">
<code class="descclassname">BoolODE.simulator.</code><code class="descname">noise</code><span class="sig-paren">(</span><em>x</em>, <em>t</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/simulator.html#noise"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.simulator.noise" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="BoolODE.simulator.simulateModel">
<code class="descclassname">BoolODE.simulator.</code><code class="descname">simulateModel</code><span class="sig-paren">(</span><em>Model</em>, <em>y0</em>, <em>parameters</em>, <em>isStochastic</em>, <em>tspan</em>, <em>seed</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/simulator.html#simulateModel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.simulator.simulateModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Call numerical integration functions, either odeint() from Scipy,
or simulator.eulersde() defined in simulator.py. By default, stochastic simulations are
carried out using simulator.eulersde.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>Model</strong> (<em>function</em>) – Function defining ODE model</li>
<li><strong>y0</strong> (<em>array</em>) – array of initial values for each state variable</li>
<li><strong>parameters</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a>) – list of parameter values to be used in simulations</li>
<li><strong>isStochastic</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a>) – User defined parameter. Default = True, can be turned off to perform ODE simulations.</li>
<li><strong>tspan</strong> (<em>ndarray</em>) – Time points to simulate</li>
<li><strong>seed</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.8)"><em>float</em></a>) – Seed to initialize random number generator</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><ul class="simple">
<li>P: Time course from numerical integration</li>
</ul>
</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">ndarray</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-BoolODE.utils">
<span id="boolode-utils-module"></span><h2>BoolODE.utils module<a class="headerlink" href="#module-BoolODE.utils" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="BoolODE.utils.checkValidInputPath">
<code class="descclassname">BoolODE.utils.</code><code class="descname">checkValidInputPath</code><span class="sig-paren">(</span><em>path</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/utils.html#checkValidInputPath"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.utils.checkValidInputPath" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns dataframe of file at path.
If path is not valid, returns empty dataframe.</p>
</dd></dl>

<dl class="function">
<dt id="BoolODE.utils.checkValidModelDefinitionPath">
<code class="descclassname">BoolODE.utils.</code><code class="descname">checkValidModelDefinitionPath</code><span class="sig-paren">(</span><em>path</em>, <em>name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/utils.html#checkValidModelDefinitionPath"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.utils.checkValidModelDefinitionPath" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="BoolODE.utils.generateInputFiles">
<code class="descclassname">BoolODE.utils.</code><code class="descname">generateInputFiles</code><span class="sig-paren">(</span><em>resultDF</em>, <em>BoolDF</em>, <em>withoutRules</em>, <em>parameterInputsDF</em>, <em>tmax</em>, <em>numcells</em>, <em>outPrefix=''</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/utils.html#generateInputFiles"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.utils.generateInputFiles" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates input files required from the Beeline pipeline</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>resultDF</strong> (<em>pandas DataFrame</em>) – The simulation output, rows are genes, columns are “cells” or timepoints</li>
<li><strong>outputfilenames</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a>) – List of filenames generated containing individual time courses</li>
<li><strong>BoolDF</strong> (<em>pandas DataFrame</em>) – Dataframe containing rules</li>
<li><strong>withoutrules</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a>) – List of nodes in input file without rules</li>
<li><strong>parameterInputsPath</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – Specifies if there are any inputs to the model</li>
<li><strong>outPrefix</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a><em> (</em><em>Optional</em><em>)</em>) – Prefix specifying target directory</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="BoolODE.utils.getRegulatorsInRule">
<code class="descclassname">BoolODE.utils.</code><code class="descname">getRegulatorsInRule</code><span class="sig-paren">(</span><em>rule</em>, <em>species</em>, <em>inputs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/utils.html#getRegulatorsInRule"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.utils.getRegulatorsInRule" title="Permalink to this definition">¶</a></dt>
<dd><p>Helper function to tokenize a rule into regulators.
Returns three lists of regulator names.
1. allreg is the list of all valid regulators
2. regulatorySpecies are other model variables that are regulators
3. inputreg are the regulators that are model inputs</p>
</dd></dl>

<dl class="function">
<dt id="BoolODE.utils.getSaneNval">
<code class="descclassname">BoolODE.utils.</code><code class="descname">getSaneNval</code><span class="sig-paren">(</span><em>size</em>, <em>lo=1.0</em>, <em>hi=10.0</em>, <em>mu=2.0</em>, <em>sig=2.0</em>, <em>identicalPars=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/utils.html#getSaneNval"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.utils.getSaneNval" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a gaussian random number which is
bounded by <cite>lo</cite> and <cite>hi</cite></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.8)"><em>int</em></a>) – number of random numbers to generate</li>
<li><strong>lo</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.8)"><em>float</em></a>) – lower bound of sample range</li>
<li><strong>hi</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.8)"><em>float</em></a>) – upper bound of sample range</li>
<li><strong>mu</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.8)"><em>float</em></a>) – Mean of Gaussian distribution to be sampled from</li>
<li><strong>sigma</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.8)"><em>float</em></a>) – Standard deviation of Gausssian distribution to be sampled from</li>
<li><strong>identicalPars</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a>) – Flag to sample single value and return a list of identical values</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><ul class="simple">
<li>K: list of sampled values</li>
</ul>
</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="BoolODE.utils.get_ss">
<code class="descclassname">BoolODE.utils.</code><code class="descname">get_ss</code><span class="sig-paren">(</span><em>P</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/utils.html#get_ss"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.utils.get_ss" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the final time point of the simulated time course</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>P</strong> (<em>np.array</em>) – 2-D array containing time course of the model</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><ul class="simple">
<li>ss : np.array containing the last entry in each time series list</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="BoolODE.utils.heavisideThreshold">
<code class="descclassname">BoolODE.utils.</code><code class="descname">heavisideThreshold</code><span class="sig-paren">(</span><em>value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/utils.html#heavisideThreshold"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.utils.heavisideThreshold" title="Permalink to this definition">¶</a></dt>
<dd><p>Decides whether the basal parameter omega_0 should 
be 1 (basal expression) or -1 (basal repression)</p>
</dd></dl>

<dl class="function">
<dt id="BoolODE.utils.loadParameterValues">
<code class="descclassname">BoolODE.utils.</code><code class="descname">loadParameterValues</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/utils.html#loadParameterValues"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.utils.loadParameterValues" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks for valid parameters in parameters.yaml file</p>
</dd></dl>

<dl class="function">
<dt id="BoolODE.utils.minmaxnorm">
<code class="descclassname">BoolODE.utils.</code><code class="descname">minmaxnorm</code><span class="sig-paren">(</span><em>X</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/utils.html#minmaxnorm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.utils.minmaxnorm" title="Permalink to this definition">¶</a></dt>
<dd><p>Scales the values in X</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>X</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.8)"><em>list</em></a>) – Input list of values to be scaled</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><ul class="simple">
<li>N : list of values scaled between min and max values in input list</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="BoolODE.utils.normalizeData">
<code class="descclassname">BoolODE.utils.</code><code class="descname">normalizeData</code><span class="sig-paren">(</span><em>P</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/utils.html#normalizeData"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.utils.normalizeData" title="Permalink to this definition">¶</a></dt>
<dd><p>Calls minmaxnorm() for each time series</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">Pnorm : list of scaled values</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="BoolODE.utils.normalizeExp">
<code class="descclassname">BoolODE.utils.</code><code class="descname">normalizeExp</code><span class="sig-paren">(</span><em>DF</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/utils.html#normalizeExp"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.utils.normalizeExp" title="Permalink to this definition">¶</a></dt>
<dd><p>Calls minmaxnorm() for each gene across all experiments</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>DF</strong> (<em>pd.DataFrame</em>) – Dataframe containing gene expression values as rows, time points as columns</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">newDF : DataFrame with values of DF scaled between min and max of DF</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="BoolODE.utils.sampleCellFromTraj">
<code class="descclassname">BoolODE.utils.</code><code class="descname">sampleCellFromTraj</code><span class="sig-paren">(</span><em>cellid</em>, <em>tspan</em>, <em>P</em>, <em>varmapper</em>, <em>sampleAt</em>, <em>genelist</em>, <em>proteinlist</em>, <em>header</em>, <em>writeProtein=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/utils.html#sampleCellFromTraj"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.utils.sampleCellFromTraj" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns pandas DataFrame with columns corresponding to 
time points and rows corresponding to genes</p>
</dd></dl>

<dl class="function">
<dt id="BoolODE.utils.sampleTimeSeries">
<code class="descclassname">BoolODE.utils.</code><code class="descname">sampleTimeSeries</code><span class="sig-paren">(</span><em>num_timepoints</em>, <em>expnum</em>, <em>tspan</em>, <em>P</em>, <em>varmapper</em>, <em>timeIndex</em>, <em>genelist</em>, <em>proteinlist</em>, <em>header</em>, <em>writeProtein=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/BoolODE/utils.html#sampleTimeSeries"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#BoolODE.utils.sampleTimeSeries" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns pandas DataFrame with columns corresponding to 
time points and rows corresponding to genes</p>
</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="BoolODE-intro.html" class="btn btn-neutral float-left" title="BoolODE" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Aditya Pratapa

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>